File: geomajas/face/geomajas-face-gwt/geomajas-gwt-client/src/main/java/org/geomajas/gwt/client/map/feature/Feature.java
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.geomajas.gwt.client.map.feature;

import java.util.HashMap;
import java.util.Map;

import org.geomajas.geometry.Coordinate;
import org.geomajas.gwt.client.gfx.Paintable;
import org.geomajas.gwt.client.gfx.PainterVisitor;
import org.geomajas.gwt.client.map.layer.VectorLayer;
import org.geomajas.gwt.client.spatial.Bbox;
import org.geomajas.gwt.client.spatial.geometry.Geometry;
import org.geomajas.gwt.client.util.GeometryConverter;
import org.geomajas.layer.feature.Attribute;
import org.geomajas.layer.feature.attribute.BooleanAttribute;
import org.geomajas.layer.feature.attribute.CurrencyAttribute;
import org.geomajas.layer.feature.attribute.DateAttribute;
import org.geomajas.layer.feature.attribute.DoubleAttribute;
import org.geomajas.layer.feature.attribute.FloatAttribute;
import org.geomajas.layer.feature.attribute.ImageUrlAttribute;
import org.geomajas.layer.feature.attribute.IntegerAttribute;
import org.geomajas.layer.feature.attribute.LongAttribute;
import org.geomajas.layer.feature.attribute.ShortAttribute;
import org.geomajas.layer.feature.attribute.StringAttribute;
import org.geomajas.layer.feature.attribute.UrlAttribute;

/**
 * <p>
 * Definition of an object belonging to a {@link VectorLayer}. Simply put, a feature is an object with a unique ID
 * within it's layer, a list of alpha-numerical attributes and a geometry.
 * </p>
 * 
 * @author Pieter De Graef
 */
public class Feature implements Paintable, Cloneable {

	/** Unique identifier */
	private String id;

	/** Reference to the layer. */
	private VectorLayer layer;

	/** Map of this feature's attributes. */
	private Map<String, Attribute> attributes;

	/** This feature's geometry. */
	private Geometry geometry;

	/** The identifier of this feature's style. */
	private int styleId;

	/** Coordinate for the label. */
	private Coordinate labelPosition;

	/** is the feature clipped ? */
	private boolean clipped;

	/**
	 * Is it allowed for the user in question to edit this feature?
	 */
	private boolean updatable;

	/**
	 * Is it allowed for the user in question to delete this feature?
	 */
	private boolean deletable;

	private String crs;

	// Constructors:

	public Feature() {
		this(null);
	}

	public Feature(VectorLayer layer) {
		this(null, layer);
	}

	public Feature(org.geomajas.layer.feature.Feature dto, VectorLayer layer) {
		this.layer = layer;
		this.attributes = new HashMap<String, Attribute>();
		this.geometry = null;
		this.styleId = 1;
		this.labelPosition = null;
		this.clipped = false;
		if (dto != null) {
			attributes = dto.getAttributes();
			id = dto.getId();
			geometry = GeometryConverter.toGwt(dto.getGeometry());
			styleId = dto.getStyleId();
			crs = dto.getCrs();
			setUpdatable(dto.isUpdatable());
			setDeletable(dto.isDeletable());
		}
	}

	// Paintable implementation:

	public void accept(PainterVisitor visitor, Bbox bounds, boolean recursive) {
		visitor.visit(this);
	}

	public String getId() {
		return id;
	}

	// Class specific functions:

	public Feature clone() {
		Feature feature = new Feature(this.layer);
		if (null != attributes) {
			feature.attributes = new HashMap<String, Attribute>(attributes);
		}
		feature.clipped = clipped;
		feature.labelPosition = labelPosition;
		feature.layer = layer;
		if (null != geometry) {
			feature.geometry = (Geometry) geometry.clone();
		}
		feature.id = id;
		feature.styleId = styleId;
		feature.crs = crs;
		feature.deletable = deletable;
		feature.updatable = updatable;
		return feature;
	}

	public Object getAttributeValue(String attributeName) {
		Attribute attribute = getAttributes().get(attributeName);
		if (attribute instanceof BooleanAttribute) {
			return ((BooleanAttribute) attribute).getValue();
		} else if (attribute instanceof CurrencyAttribute) {
			return ((CurrencyAttribute) attribute).getValue();
		} else if (attribute instanceof DateAttribute) {
			return ((DateAttribute) attribute).getValue();
		} else if (attribute instanceof DoubleAttribute) {
			return ((DoubleAttribute) attribute).getValue();
		} else if (attribute instanceof FloatAttribute) {
			return ((FloatAttribute) attribute).getValue();
		} else if (attribute instanceof ImageUrlAttribute) {
			return ((ImageUrlAttribute) attribute).getValue();
		} else if (attribute instanceof IntegerAttribute) {
			return ((IntegerAttribute) attribute).getValue();
		} else if (attribute instanceof LongAttribute) {
			return ((LongAttribute) attribute).getValue();
		} else if (attribute instanceof ShortAttribute) {
			return ((ShortAttribute) attribute).getValue();
		} else if (attribute instanceof StringAttribute) {
			return ((StringAttribute) attribute).getValue();
		} else if (attribute instanceof UrlAttribute) {
			return ((UrlAttribute) attribute).getValue();
		}
		return null;
	}

	/**
	 * Transform this object into a DTO feature.
	 * 
	 * @return dto for this feature
	 */
	public org.geomajas.layer.feature.Feature toDto() {
		org.geomajas.layer.feature.Feature dto = new org.geomajas.layer.feature.Feature();
		dto.setAttributes(attributes);
		dto.setClipped(clipped);
		dto.setId(id);
		dto.setGeometry(GeometryConverter.toDto(geometry));
		dto.setCrs(crs);
		return dto;
	}

	public String getLabel() {
		String attributeName = layer.getLayerInfo().getNamedStyleInfo().getLabelStyle().getLabelAttributeName();
		Object attributeValue = getAttributeValue(attributeName);
		return attributeValue == null ? "null" : attributeValue.toString();
	}

	// Getters and setters:

	/**
	 * Crs as (optionally) set by the backend.
	 *
	 * @return crs for this feature
	 */
	public String getCrs() {
		return crs;
	}

	/**
	 * Get the attributes map, null when it needs to be lazy loaded.
	 *
	 * @return attributes map
	 * @throws IllegalStateException attributes not present because of lazy loading
	 */
	public Map<String, Attribute> getAttributes() throws IllegalStateException {
		if (null == attributes) {
			throw new IllegalStateException("Attributes not available, use LazyLoader.");
		}
		return attributes;
	}

	/**
	 * Set the attributes map.
	 *
	 * @param attributes attributes map
	 */
	public void setAttributes(Map<String, Attribute> attributes) {
		this.attributes = attributes;
	}

	/**
	 * Check whether the attributes are already available or should be lazy loaded.
	 *
	 * @return true when attributes are available
	 */
	public boolean isAttributesLoaded() {
		return attributes != null;
	}

	/**
	 * Get the feature's geometry, , null when it needs to be lazy loaded.
	 *
	 * @return geometry
	 * @throws IllegalStateException attributes not present because of lazy loading
	 */
	public Geometry getGeometry() throws IllegalStateException {
		if (null == geometry) {
			throw new IllegalStateException("Geometry not available, use LazyLoader.");
		}
		return geometry;
	}

	/**
	 * Set the geometry.
	 *
	 * @param geometry geometry
	 */
	public void setGeometry(Geometry geometry) {
		this.geometry = geometry;
	}

	/**
	 * Check whether the geometry is already available or should be lazy loaded.
	 *
	 * @return true when geometry are available
	 */
	public boolean isGeometryLoaded() {
		return geometry != null;
	}

	public int getStyleId() {
		return styleId;
	}

	public void setStyleId(int styleId) {
		this.styleId = styleId;
	}

	public boolean isSelected() {
		return layer.isFeatureSelected(getId());
	}

	public VectorLayer getLayer() {
		return layer;
	}

	/**
	 * Is the logged in user allowed to edit this feature?
	 * 
	 * @return true when edit/update is allowed for this feature
	 */
	public boolean isUpdatable() {
		return updatable;
	}

	/**
	 * Set whether the logged in user is allowed to edit/update this feature.
	 * 
	 * @param editable
	 *            true when edit/update is allowed for this feature
	 */
	public void setUpdatable(boolean editable) {
		this.updatable = editable;
	}

	/**
	 * Is the logged in user allowed to delete this feature?
	 * 
	 * @return true when delete is allowed for this feature
	 */
	public boolean isDeletable() {
		return deletable;
	}

	/**
	 * Set whether the logged in user is allowed to delete this feature.
	 * 
	 * @param deletable
	 *            true when deleting this feature is allowed
	 */
	public void setDeletable(boolean deletable) {
		this.deletable = deletable;
	}

}

File: geomajas/face/geomajas-face-dojo/geomajas-dojo-client/src/main/resources/geomajas/map/Feature.js
dojo.provide("geomajas.map.Feature");
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
dojo.require("geomajas.gfx.PainterVisitable");
dojo.require("geomajas.gfx.PainterVisitor");

dojo.require("geomajas.map.attributes.proxy.ProxyAttributeMap");
dojo.require("geomajas.map.attributes.proxy.ProxyGeometry");

dojo.declare("Feature", PainterVisitable, {

	/**
	 * @fileoverview Element of a layer (vectorlayer).
	 * @class The objects in layers are called Features. And these features in
	 * turn are a composition of a Geometry (for the geographical representation)
	 * and attributes (for alpha-numeric representation). Since features need
	 * to be painted on the map, this class must implement the PainterVisitable
	 * interface.
	 * @author Pieter De Graef & Jan De Moerloose
	 *
	 * @constructor
	 * @extends PainterVisitable
	 */
	constructor : function () {
		/** Unique identifier, over layers, and even maps. */
		this.id = null;

		/** Reference to the layer. */
		this.layer = null;

		/** Map of this feature's attributes. */
		this.attributes = {};

		/** This feature's geometry. */
		this.geometry = null;
		
		/** This feature's path attribute */
		this.path = null;

		/** The identifier of this feature's style. */
		this.styleId = null;

		/** Coordinate for the label. */
		this.labelPosition = null;
		
		/** is the feature clipped ? */
		this.clipped = false;
		
		this.editable = false;
	},

	/**
	 * Implemantation of the PainterVisitable interface.
	 * Accept visitor visiting.
	 */
	accept : function (/*PainterVisitor*/visitor, /*Bbox*/ bbox, recursive) {
		visitor.visit(this);
	},

	/**
	 * Create a clone of this feature.
	 */
	clone : function () {
		var f = new Feature();
		f.setId(this.id);
		f.setLayer(this.getLayer());
		if (this.geometry != null) {
			f.setGeometry(this.geometry.clone()); // proxy or not; works on both
		}
		if (this.attributes != null) {
			if (this.attributes instanceof ProxyAttributeMap || 
					this.attributes.declaredClass == "ProxyAttributeMap") {
				f.setAttributes(this.attributes.clone());
			} else {
				var attributes = {};
				for (attr in this.attributes) {
					attributes[attr] = this.attributes[attr];
				}
				f.setAttributes(attributes);
			}
		}
		if (this.labelPosition != null) {
			f.setLabelPosition(this.labelPosition.clone());
		}
		f.setClipped(this.clipped);
		f.setPath(this.path);
		return f;
	},



	//-------------------------------------------------------------------------
	// JSON serialization:
	//-------------------------------------------------------------------------

	/**
	 * Prepare this object for JSON serialization.
	 * @returns A recursive hashmap of name-value pairs.
	 */
	toJSON : function () {
        var jsonAttr = {};
        var temp = this.getAttributes(); // make sure they are actually fetched!
        for (attr in temp) {
            // special treatment for dates !
            if(temp[attr] instanceof Date){
                jsonAttr[attr]  = {
                      "javaClass": "java.util.Date",
                      "time": temp[attr].getTime()
                }
            } else if(temp[attr] instanceof Array && temp[attr].length == 0){
                jsonAttr[attr]  = {
                      "javaClass": "java.util.ArrayList",
                      "list": []
                }
            }
            else {
                jsonAttr[attr] = temp[attr];
            }
        }
        return {
			javaClass : "org.geomajas.layer.feature.Feature",
			id : this.getId(),
			geometry : this.getGeometry(),
		    attributes : {
                "javaClass": "java.util.HashMap",
                "map": jsonAttr
            }
        };
	},

	/**
	 * Accepts a JSON string, and fills in all fields of this feature object.
	 * @param json A recursive hashmap of name-value pairs.
	 */
	fromJSON : function (json) {
		this.setId(json.id);
		this.setClipped(json.clipped);
		this.setPath(json.path);

		if (json.attributes) {
			this.setAttributes(json.attributes.map);
		} else {
			this.attributes = new ProxyAttributeMap(this.getLayer().getId(), this.getLocalId());
		}
		
		this.styleId = json.styleId;
		if (json.geometry) {
			var deser = new GeometryDeserializer();
			var geometry = deser.createGeometryFromJSON (json.geometry);
			this.setGeometry(geometry);
		} else {
			this.setGeometry(new ProxyGeometry(this.getLayer().getId(), this.getLocalId()));
		}
		if (json.labelPosition) {
			this.setLabelPosition(new Coordinate(json.labelPosition.x, json.labelPosition.y));
		}
		if (json.editable) {
			this.editable = json.editable;
		}
		// make sure selectionstore stays uptodate
		var selStore = this.getLayer().getSelectionStore();
		if (selStore.contains(this.getLocalId())) {
			selStore.remove(this.getLocalId());
			selStore.add(this.getLocalId(),this);
		}
	},



	//-------------------------------------------------------------------------
	// ID fetching
	//-------------------------------------------------------------------------

	getId : function () {
		return this.id;
	},

	setId : function (id) {
		this.id = id;
	},

	/**
	 * Returns the feature's own simple ID.
	 */
	getLocalId : function () {
		if (this.id != null) {
			var pos = this.id.lastIndexOf(".");
			if (pos >= 0) {
				return this.id.substring(pos + 1);
			}
		}
		return null;
	},

	/**
	 * Returns the DOM id of the selected feature
 	 */
	getSelectionId : function () {
		return this.getLayer().getId() + ".selection." + this.getLocalId();
	},


	//-------------------------------------------------------------------------
	// Attribute handling:
	//-------------------------------------------------------------------------

	/**
	 * Get the value the attribute corresponding to the given attribute name.
	 * In case of an association, the return object can be quite complex.
	 * @param name The name of the attribute.
	 * @return Returns the primitive value or null.
	 */
	getAttributeValue : function (name) {
		this._checkAttributes();
		return this.attributes[name].value;
	},

	/**
	 * Sets a new value for one of the attributes. In case of an association,
	 * make sure you pass a complex object as value, because this function
	 * tries to set both it's identifier, as it's label value.
	 * @param name The attribute's name.
	 * @param value The value for the attribute. Complex in case of an
	 *              association!
	 */
	setAttributeValue : function (name, value) {
		this._checkAttributes();
		if(this.getLayer()) {
			var attribute = this.getLayer().getFeatureType().getAttributeByName(name);
			if(attribute.isAssociation()){
				// TODO !!!
			}  else {
				this.attributes[name].value = value;
			}
			/**
			if (attribute != null && attribute.isAssociation() && value != null && attribute.getType() == "many-to-one") {
				var association = this.attributes[name];
				if (association == null) { // if no value yet...
					association = {javaClass : attribute.object.name };
				}
				var identifierName = attribute.getAssociationIdentifierName();
				association[identifierName] = value[identifierName];

				var associationName = attribute.getAssociationAttributeName();
				association[associationName] = value[associationName];

				this.attributes[name] = association; // in case association was null...
			} else {
				this.attributes[name] = value;
			}
			*/
		} else {
			this.attributes[name] = value;
		}
	},

	/**
	 * Return this feature's label. It uses the preconfigured labelattribute to
	 * do so. If no label is found, an empty string is returned.
	 */
	getLabel : function () {
		if (this.layer) {
			var attrName = this.layer.getLabelStyle().getLabelAttributeName();
			var value = this.getAttributeValue(attrName);
			if (value != null) {
				return value;
			}
		}
		return "";
	},



	//-------------------------------------------------------------------------
	// Getters and setters:
	//-------------------------------------------------------------------------

	getLayer : function () {
		return this.layer;
	},
	
	setLayer : function (layer) {
		this.layer = layer;
	},

	getAttributes : function () {
		this._checkAttributes();
		return this.attributes;
	},

	setAttributes : function (attributes) {
		this.attributes = this._fixDates(attributes);
	},

	getGeometry : function () {
		if (this.geometry instanceof ProxyGeometry) {
			var temp = this.geometry.getValue();
			this.geometry = temp;
		}
		return this.geometry;
	},

	setGeometry : function (geometry) {
		this.geometry = geometry;
	},

	getPath : function () {
		return this.path;
	},

	setPath : function (path) {
		this.path = path;
	},

	getStyle : function () {
		return this.getLayer().getStyleByIndex(this.styleId).getStyle();
	},

	setStyleId : function (styleId) {
		this.styleId = styleId;
	},

	getStyleId : function () {
		return this.styleId;
	},

	getLabelPosition : function () {
		return this.labelPosition;
	},

	setLabelPosition : function (labelPosition) {
		this.labelPosition = labelPosition;
	},

	isSelected : function () {
		if (this.layer && this.layer instanceof VectorLayer) {
//			var other = this.layer.getSelectionStore().item(this.getLocalId());
//			if (other) {
//				this.geometry = other.getGeometry();
//				return true;
//			}
			return this.layer.getSelectionStore().contains(this.getLocalId());
		}
		return false;
	},

	isClipped : function () {
		return this.clipped;
	},

	setClipped : function (clipped) {
		this.clipped = clipped;
	},

	isEditable : function () {
		return this.editable;
	},

	setEditable : function (editable) {
		this.editable = editable;
	},

	_simplifyFromJSON : function (json) {
		var result = [];
		for(var key in json){
			if(key != "id" && key !="javaClass"){
				result.push(key+"="+json[key]);
			}
		}
		return "{"+result.join(",")+"}";
	},
	
	_checkAttributes : function () {
		if (this.attributes != null && (this.attributes instanceof ProxyAttributeMap || 
				this.attributes.declaredClass == "ProxyAttributeMap")) {
			var json = this.attributes.getValue();
			this.setAttributes(json);
		}
	},

	_fixDates : function (obj) {
		// Dates aren't parsed by Dojo JSON parser (which uses eval).
		// We recursively replace java dates by javascript dates
		if(obj && (obj instanceof Object || obj instanceof Array)){
			for (var i in obj) {
				var arrVal = obj[i];
				if (this._isDate(arrVal)) {
					obj[i] = this._toDate(arrVal);
				} else if(arrVal){
					obj[i] = this._fixDates(arrVal);
				}
			}
		}
		return obj;
	},

	_toDate : function(obj) {
		return new Date(obj.time);
	},

	_isDate :  function(obj) {
	   return (obj != null && obj.javaClass && (obj.javaClass == "java.sql.Timestamp" || obj.javaClass == "java.util.Date"));
	}

});
